pkg/trace/api/telemetry.go

```diff
@@ -79,164 +90,231 @@ func (r *HTTPReceiver) telemetryProxyHandler() http.Handler {
-			req.Header.Set("DD-Agent-Install-Type", installSignature.InstallType)
-			req.Header.Set("DD-Agent-Install-Time", strconv.FormatInt(installSignature.InstallTime, 10))
-		}
-		if arn, ok := r.conf.GlobalTags[functionARNKeyTag]; ok {
-			req.Header.Set(cloudProviderHeader, string(aws))
-			req.Header.Set(cloudResourceTypeHeader, string(awsLambda))
-			req.Header.Set(cloudResourceIdentifierHeader, arn)
-		} else if taskArn, ok := extractFargateTask(containerTags); ok {
-			req.Header.Set(cloudProviderHeader, string(aws))
-			req.Header.Set(cloudResourceTypeHeader, string(awsFargate))
-			req.Header.Set(cloudResourceIdentifierHeader, taskArn)
+// Stop waits for up to 1s to end all telemetry forwarded requests.
+func (f *TelemetryForwarder) Stop() {
+	done := make(chan any)
+	go func() {
+		f.inflightWaiter.Wait()
+		close(done)
+	}()
+	select {
+	case <-done:
+	// Give a max 1s timeout to wait for all requests to end
+	case <-time.After(1 * time.Second):
+	}
+}
+
+func (f *TelemetryForwarder) startRequest(size int64) (accepted bool) {
+	for {
+		inflight := f.inflightCount.Load()
+		newInflight := inflight + size
+		if newInflight >= f.maxInflightBytes {
+			return false
 		}
-		if origin, ok := r.conf.GlobalTags[originTag]; ok {
-			switch origin {
-			case "cloudrun":
-				req.Header.Set(cloudProviderHeader, string(gcp))
-				req.Header.Set(cloudResourceTypeHeader, string(cloudRun))
-				if serviceName, found := r.conf.GlobalTags["service_name"]; found {
-					req.Header.Set(cloudResourceIdentifierHeader, serviceName)
-				}
-			case "appservice":
-				req.Header.Set(cloudProviderHeader, string(azure))
-				req.Header.Set(cloudResourceTypeHeader, string(azureAppService))
-				if appName, found := r.conf.GlobalTags["app_name"]; found {
-					req.Header.Set(cloudResourceIdentifierHeader, appName)
-				}
-			case "containerapp":
-				req.Header.Set(cloudProviderHeader, string(azure))
-				req.Header.Set(cloudResourceTypeHeader, string(azureContainerApp))
-				if appName, found := r.conf.GlobalTags["app_name"]; found {
-					req.Header.Set(cloudResourceIdentifierHeader, appName)
-				}
-			}
+		if f.inflightCount.CompareAndSwap(inflight, newInflight) {
+			return true
 		}
 	}
-	return &httputil.ReverseProxy{
-		Director:  director,
-		ErrorLog:  logger,
-		Transport: &transport,
-	}
 }
 
-func extractFargateTask(containerTags string) (string, bool) {
-	return extractTag(containerTags, "task_arn")
+func (f *TelemetryForwarder) endRequest(req forwardedRequest) {
+	f.inflightCount.Add(-int64(len(req.body)))
 }
 
-func extractTag(tags string, name string) (string, bool) {
-	leftoverTags := tags
-	for {
-		if leftoverTags == "" {
-			return "", false
+func (f *TelemetryForwarder) forwardTelemetryAsynchronously(req forwardedRequest) {
+	f.inflightWaiter.Add(1)
+	go func() {
+		defer f.inflightWaiter.Done()
+		defer f.endRequest(req)
+
+		f.setRequestHeader(req.req)
+		f.forwardTelemetry(context.Background(), req)
```

Comment:
should probably be the context in the request?

Reply:
I don't think so, we want the context lifecycle to be decoupled from the connection

https://pkg.go.dev/net/http#Request.Context
> For incoming server requests, the context is canceled when the client's connection closes, the request is canceled (with HTTP/2), or when the ServeHTTP method returns.